File: design-patterns/java/State-DesignPattern/src/WithoutStatePattern/ATMMachine.java
package WithoutStatePattern;

public class ATMMachine {

    private String state; //Idle, Has card, Authenticated, Dispense
    private int balance;

    public ATMMachine(int balance) {
        this.balance = balance;
        this.state = "IDLE";
    }

    public void insertCard() {

        if (state.equals("IDLE")) {
            System.out.println("Card inserted");
            state = "HAS_CARD";
        }
        else {
            System.out.println("Card already inserted");
        }
    }

    public void enterPin(int pin) {

        if (state.equals("HAS_CARD")) {

            if (pin == 1234) {
                System.out.println("PIN correct");
                state = "AUTHENTICATED";
            } else {
                System.out.println("Wrong PIN");
                state = "IDLE";
            }

        } else if (state.equals("IDLE")) {
            System.out.println("Insert card first");
        }
        else {
            System.out.println("Already authenticated");
        }
    }

    public void withdrawCash(int amount) {

        if (state.equals("AUTHENTICATED")) {

            if (balance >= amount) {
                System.out.println("Dispensing cash: " + amount);
                balance -= amount;
                state = "IDLE";
            } else {
                System.out.println("Insufficient balance");
            }

        } else if (state.equals("HAS_CARD")) {
            System.out.println("Enter PIN first");
        }
        else {
            System.out.println("Insert card first");
        }
    }

    public void ejectCard() {

        if (state.equals("HAS_CARD") || state.equals("AUTHENTICATED")) {
            System.out.println("Card ejected");
            state = "IDLE";
        }
        else {
            System.out.println("No card inserted");
        }
    }
}

File: design-patterns/python/State-DesignPattern/src/without_state_pattern/atm_machine.py
"""ATM without State pattern - logic in if/else based on state string."""
from typing import Literal


class ATMMachine:
    def __init__(self, balance: int) -> None:
        self._balance = balance
        self._state: Literal["IDLE", "HAS_CARD", "AUTHENTICATED"] = "IDLE"

    def insert_card(self) -> None:
        if self._state == "IDLE":
            print("Card inserted")
            self._state = "HAS_CARD"
        else:
            print("Card already inserted")

    def enter_pin(self, pin: int) -> None:
        if self._state == "HAS_CARD":
            if pin == 1234:
                print("PIN correct")
                self._state = "AUTHENTICATED"
            else:
                print("Wrong PIN")
                self._state = "IDLE"
        elif self._state == "IDLE":
            print("Insert card first")
        else:
            print("Already authenticated")

    def withdraw_cash(self, amount: int) -> None:
        if self._state == "AUTHENTICATED":
            if self._balance >= amount:
                print(f"Dispensing cash: {amount}")
                self._balance -= amount
                self._state = "IDLE"
            else:
                print("Insufficient balance")
        elif self._state == "HAS_CARD":
            print("Enter PIN first")
        else:
            print("Insert card first")

    def eject_card(self) -> None:
        if self._state in ("HAS_CARD", "AUTHENTICATED"):
            print("Card ejected")
            self._state = "IDLE"
        else:
            print("No card inserted")
